BanchoBot/functions/tournaments/matchup/runMatchup.ts

```diff
@@ -0,0 +1,328 @@
+            log(matchup, `Error loading beatmap: ${ex}`);
+        }
+    });
+
+    // Player left event
+    mpLobby.on("playerLeft", async (player) => {
+        log(matchup, `Player ${player.user.username} left the lobby`);
+
+        if (
+            mpLobby.playing &&
+            playersInLobby.some(p => p.user.id === player.user.id)
+        ) {
+            const team = matchup.teams!.find(team => team.members.some(m => m.osu.userID === player.user.id.toString()));
+            if (
+                (team &&
+                    (
+                        aborts.get(team.ID) === undefined ||
+                        !matchup.stage!.tournament.teamAbortLimit ||
+                        aborts.get(team.ID)! <= matchup.stage!.tournament.teamAbortLimit
+                    )
+                ) && (
+                    matchStart &&
+                    Date.now() - matchStart.getTime() < (matchup.stage!.tournament.abortThreshold || 30) * 1000
+                )
+            ) {
+                await Promise.all([
+                    mpLobby.abortMatch(),
+                    mpChannel.sendMessage(`${player.user.username} left the match so we're aborting`),
+                ]);
+                aborts.set(team.ID, (aborts.get(team.ID) || 0) + 1);
+            }
+        }
+
+        playersInLobby = playersInLobby.filter(p => p.user.id !== player.user.id);
+    });
+
+    // All players ready event
+    mpLobby.on("allPlayersReady", async () => {
+        if (mapsPlayed.some(m => m.beatmap!.ID === mpLobby.beatmapId))
+            return;
+
+        if (!allPlayersInLobby(matchup, playersInLobby)) {
+            await mpChannel.sendMessage("bruh just cuz ur all ready doesnt mean anything if not enough players are in each team to start yet hrur y up");
+            return;
+        }
+
+        if (!consistentPlayers(mpLobby, playersPlaying)) {
+            await mpChannel.sendMessage("u are so sneaky!!1!! Now get the same players that were in the map before the abort in here or else .");
+            return;
+        }
+
+        const slotMod = pools.flatMap(p => p.slots).find(s => s.maps.some(map => map.beatmap!.ID === mpLobby.beatmapId));
+        if (!slotMod) {
+            await mpChannel.sendMessage("bruh this map isnt in any of the pools??? COINTACT CORSACE IMMEDIATELY");
+            return;
+        }
+        if (!hasAllMods(mpLobby, slotMod)) {
+            await mpChannel.sendMessage(`SOMEEONEEE HAS THE WRONG MODS ON . Allowed mods for this slot are ${getStageMods(slotMod.allowedMods).map(m => `${m.longMod} (${m.shortMod})`).join(", ")}`);
+            return;
+        }
+
+        log(matchup, "All players readied up for the next map");
+        await mpLobby.startMatch(5);
+
+        const abort = async () => {
+            await mpLobby.abortTimer();
+            await mpLobby.removeListener("playerLeft", abort);
+        };
+        mpLobby.on("playerLeft", abort);
```

Comment:
this is not unset anywhere i think. also i highly recommend against dynamically adding/removing listeners during the match to prevent this exact kind of oversight. basically, only define static listeners here on match creation.